krun/platform.py

```diff
@@ -132,26 +133,63 @@ def _collect_dmesg_lines(self):
     def _timestamp_to_str(self, lt):
         return time.strftime(ABS_TIME_FORMAT, lt)
 
+    def get_allowed_dmesg_patterns(self):
+        return []
+
+    def filter_new_dmesg_line(self, line, patterns):
+        for p in patterns:
+            if p.match(line):
+                debug("Allowed dmesg change: '%s'" % line)
+                return True  # allowed change of dmesg
+        return False
+
     def check_dmesg_for_changes(self):
         new_dmesg_time = localtime()
         new_dmesg = self._collect_dmesg_lines()
+        patterns = self.get_allowed_dmesg_patterns()
 
-        old_fn = self._timestamp_to_str(self.last_dmesg_time)
-        new_fn = self._timestamp_to_str(new_dmesg_time)
-        lines = [x for x in difflib.unified_diff(
-            self.last_dmesg, new_dmesg, old_fn, new_fn, lineterm="")]
+        rv = self._check_dmesg_for_changes(patterns,
+                                           self.last_dmesg,
+                                           self.last_dmesg_time, new_dmesg,
+                                           new_dmesg_time)
 
-        if lines:
-            # dmesg changed!
-            diff = "\n".join(lines)
-            warn_s = "dmesg seems to have changed! Diff follows:\n" + diff
-            log_and_mail(self.mailer, warn, "dmesg changed", warn_s)
+        self.last_dmesg = new_dmesg
+        self.last_dmesg_time = new_dmesg_time
 
-            self.last_dmesg = new_dmesg
-            self.last_dmesg_time = new_dmesg_time
+        return rv
 
-            return True  # i.e. a (potential) error occurred
-        return False
+    def _check_dmesg_for_changes(self, patterns, last_dmesg, last_dmesg_time,
+                                 new_dmesg, new_dmesg_time):
+        differ = difflib.Differ()
+        delta = differ.compare(last_dmesg, new_dmesg)
+        problem_lines = []
+
+        for rec in delta:
+            code = rec[0:2]
+            line = rec[2:]
+
+            if code == "  ":  # in both
+                pass
+            elif code == "+ ":  #  new line in dmesg
+                if not self.filter_new_dmesg_line(line, patterns):
+                    # this is a new line which we did not expect
+                    problem_lines.append(rec)
+            else:
+                assert False  # The other codes should not arise
+
+        rv = False
+        if problem_lines:
+            # dmesg changed! Emit something which looks a bit like a diff.
+            old_fn = "old_dmesg   %s" % self._timestamp_to_str(last_dmesg_time)
+            new_fn = "new_dmesg   %s" % self._timestamp_to_str(new_dmesg_time)
+
+            warn_s = ("dmesg seems to have (unexpectedly) changed.\n"
```

Comment:
Yes, but the user will know that, if reported, they need to look at things. No need to go overboard in the explanation IMHO.